src/namedtuples.jl

```diff
@@ -28,47 +35,62 @@ function Base.iterate(rows::NamedTupleIterator{NT}, st=()) where {NT <: NamedTup
         x = iterate(rows.x, st...)
         x === nothing && return nothing
         row, st = x
-        return NT(Tuple(getproperty(row, fieldtype(NT, i), i, nm) for i = 1:fieldcount(NT))), (st,)
+        if T === nothing
+            return NamedTuple{names}(Tuple(getproperty(row, nm) for i = 1:length(names))), (st,)
+        else
+            return NamedTuple{names, T}(Tuple(getproperty(row, fieldtype(T, i), i, nm) for i = 1:fieldcount(T))), (st,)
+        end
     end
 end
 
-namedtupleiterator(::Type{T}, sch, rows) where {T <: NamedTuple} = rows
-namedtupleiterator(T, sch, rows) = NamedTupleIterator(sch, rows)
+namedtupleiterator(::Type{T}, rows::S) where {T <: NamedTuple, S} = rows
+namedtupleiterator(::Type{T}, rows::S) where {T, S} = NamedTupleIterator{typeof(schema(rows)), S}(rows)
 
-function rowtable(itr)
+function rowtable(itr::T) where {T}
+    istable(T) || throw(ArgumentError("Vector of NamedTuples requires a table input"))
     r = rows(itr)
-    return collect(namedtupleiterator(eltype(r), schema(itr), r))
+    return collect(namedtupleiterator(eltype(r), r))
 end
 
 # NamedTuple of Vectors
 const ColumnTable = NamedTuple{names, T} where {names, T <: NTuple{N, AbstractVector{S} where S}} where {N}
 rowcount(c::ColumnTable) = length(c) == 0 ? 0 : length(c[1])
 
 # interface implementation
-AccessStyle(::Type{<:ColumnTable}) = ColumnAccess()
+istable(::Type{<:ColumnTable}) = true
+columnaccess(::Type{<:ColumnTable}) = true
+columns(x::ColumnTable) = x
+schema(x::T) where {T <: ColumnTable} = Schema(names(T), types(T))
 
 _eltype(::Type{A}) where {A <: AbstractVector{T}} where {T} = T
-schema(ct::T) where {T <: ColumnTable} = schema(T)
-Base.@pure function schema(::Type{NT}) where {NT <: NamedTuple{names, T}} where {names, T <: NTuple{N, AbstractVector{S} where S}} where {N}
-    TT = Tuple{Any[ _eltype(fieldtype(NT, i)) for i = 1:fieldcount(NT) ]...}
-    return NamedTuple{names, TT}
+Base.@pure function types(::Type{NT}) where {NT <: NamedTuple{names, T}} where {names, T <: NTuple{N, AbstractVector{S} where S}} where {N}
+    return Tuple{Any[ _eltype(fieldtype(NT, i)) for i = 1:fieldcount(NT) ]...}
```

Comment:
Use a generator instead of an array comprehension to avoid an allocation? Could also use `ntuple`, which often gives cleaner code IIRC.